html/semantics/forms/form-submission-0/constructing-form-data-set.html

```diff
@@ -158,4 +174,48 @@
   assert_equals(formDataInEvent.get('n1'), 'v1');
   assert_false(formDataInEvent.has('n2'));
 }, 'The constructed FormData object should not contain an entry for the submit button that was used to submit the form.');
+
+test(() => {
+  let form = populateForm('<input name=n1 value=v1><button type=submit name=n2 value=v2></button><input name=n3 value=v3>');
+  let submitter = form.querySelector('button[type=submit]');
+  let formData = new FormData(form, submitter);
+  let entries = [...formData.entries()];
+
+  assert_deep_equals(entries, [["n1", "v1"], ["n2", "v2"], ["n3", "v3"]]);
+}, 'The constructed FormData object should contain an in-tree-order entry for an explicit submit button submitter');
+
+test(() => {
+  let form = populateForm('<input name=n1 value=v1><input type=image name=n2 src="1x1-green.png"></button><input name=n3 value=v3>');
+  let submitter = form.querySelector('input[type=image]');
+  form.addEventListener("submit", (event) => event.preventDefault())
+  submitter.click();
+  let formData = new FormData(form, submitter);
+  let entries = [...formData.entries()];
+
+  assert_deep_equals(entries, [["n1", "v1"], ["n2.x", "0"], ["n2.y", "0"], ["n3", "v3"]]);
+}, 'The constructed FormData object should contain in-tree-order entries for an activated Image Button submitter');
+
+test(() => {
+  let form = populateForm('<input name=n1 value=v1><input type=image name=n2 src="1x1-green.png"></button><input name=n3 value=v3>');
+  let submitter = form.querySelector('input[type=image]');
+  let formData = new FormData(form, submitter);
+  let entries = [...formData.entries()];
+
+  assert_deep_equals(entries, [["n1", "v1"], ["n3", "v3"]]);
+}, 'The constructed FormData object should not contain entries for an unactivated Image Button submitter');
```

Comment:
I believe this test is correct based on my reading of [the spec](https://html.spec.whatwg.org/multipage/form-control-infrastructure.html#constructing-the-form-data-set), i.e. `before this algorithm was invoked the user indicated a coordinate` would imply that no entries should be added in the negative case. But it could also be argued that that is undefined behavior, since it doesn't explicitly `continue` in the prose.

In looking at implementations of the algorithm, it seems that Blink and WebKit both short circuit if the control wasn't activated, while Gecko will default the coordinate to 0:0.

Would y'all recommend leaving this test in (and getting that fixed in Gecko), or removing the test and considering it undefined behavior, since the spec is slightly unclear?

As far as fixing Gecko to check the activated-ness, my hunch is that should be pretty safe/straightforward, since in normal submissions the submitter would have activated. It's only with the new FormData submitter parameter that there might be an unactivated submitter.

Reply:
I think 3 makes the most sense here. Thanks for catching that!

cc @domenic @tkent-google

Reply:
https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3Cform%20action%3D%2Futilities%2Fcgi%2Ftest-tools%2Fecho%3E%0A%20%3Cinput%20type%3Dimage%3E%0A%3C%2Fform%3E%0A%3Cscript%3E%0Aonload%20%3D%20()%20%3D%3E%20document.querySelector(%22input%22).click()%0A%3C%2Fscript%3E also shows this is a clear bug in HTML. I'm surprised we didn't catch this before.

Reply:
So @jenseng as for next steps:

1. Leave the test in, but change the pass condition.
2. File an issue against or directly PR HTML.

Given my test above this is a (long overdue given how fundamental the algorithm is) bug fix.

Reply:
@annevk sounds good, I'll tweak the test accordingly and file a separate issue/PR.

One slight clarification though, does your test actually demonstrate a bug in the spec? The `click()` method [should trigger the activation behavior](https://html.spec.whatwg.org/#activation), and when an Image Button is activated it [sets the coordinate to `(0,0)`](https://html.spec.whatwg.org/#image-button-state-(type=image):input-activation-behavior) if the user didn't explicitly pick one.

I think the ambiguity/discrepancy is just for Image Button submitters that were not explicitly activated (e.g. when using `aForm.requestSubmit(aSubmitter)` or the proposed `new FormData(aForm, aSubmitter)`. Let me know If I'm missing something.

Reply:
Oh lovely, more action-at-a-distance. You're quite right.

https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3Cform%20action%3D%2Futilities%2Fcgi%2Ftest-tools%2Fecho%3E%0A%20%3Cinput%20type%3Dimage%3E%0A%3C%2Fform%3E%0A%3Cscript%3E%0Aonload%20%3D%20()%20%3D%3E%20document.querySelector(%22form%22).requestSubmit(document.querySelector(%22input%22))%0A%3C%2Fscript%3E (using `requestSubmit()`) demonstrates it better.

I guess once we have the inline defaulting we should simplify that activation behavior language at the same time.